pkg/registry/apis/query/queryschema/oas_helper.go

```diff
@@ -0,0 +1,235 @@
+package queryschema
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/kube-openapi/pkg/spec3"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+
+	"github.com/grafana/grafana-plugin-sdk-go/experimental/schemabuilder"
+
+	data "github.com/grafana/grafana-plugin-sdk-go/experimental/apis/data/v0alpha1"
+
+	query "github.com/grafana/grafana/pkg/apis/query/v0alpha1"
+	"github.com/grafana/grafana/pkg/plugins"
+)
+
+const QueryRequestSchemaKey = "QueryRequestSchema"
+
+// const QueryPayloadSchemaKey = "QueryPayloadSchema"
+// const QuerySaveModelSchemaKey = "QuerySaveModelSchema"
+
+type OASQueryOptions struct {
+	Swagger    *spec3.OpenAPI
+	PluginJSON *plugins.JSONData
+	QueryTypes *query.QueryTypeDefinitionList
+
+	Root             string
+	QueryPath        string // eg "namespaces/{namespace}/query/{name}"
+	QueryDescription string
+	QueryExamples    map[string]*spec3.Example
+}
+
+func AddQueriesToOpenAPI(options OASQueryOptions) error {
```

Comment:
This moves the logic that was duplicated in the query and datasource apiservers -- it now produces a more valid response (still a few issues in converting jsonschema to OpenAPI, but at least it is in one place now!)